core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    private final StreamMessage<T> source;
+    private final EventExecutor executor;
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+    private final AtomicBoolean subscribed = new AtomicBoolean();
+    private volatile boolean closed;
```

Comment:
Since we assume one thread reads this `InputStream`, can we use `boolean subscribed`?
```suggestion
    private boolean subscribed;
    private boolean closed;
```

Reply:
✅ updated.